pkg/metallb/frr.go

```diff
@@ -0,0 +1,618 @@
+	}
+
+	// Validate CIDR prefixes
+	for _, prefix := range prefixes {
+		if _, _, err := net.ParseCIDR(prefix); err != nil {
+			glog.V(100).Infof("the frrConfiguration prefix %s is not a valid CIDR", prefix)
+			builder.errorMsg = fmt.Sprintf("The prefix %s is not a valid CIDR", prefix)
+		}
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	// Prepare the list of allowed prefixes
+	allowedPrefixes := make([]frrtypes.PrefixSelector, len(prefixes))
+
+	for i, prefix := range prefixes {
+		glog.V(100).Infof("AllowMode 'filtered' receives only routes within the prefix %v", prefix)
+		allowedPrefixes[i] = frrtypes.PrefixSelector{Prefix: prefix}
+	}
+
+	// Update the specific neighbor's ToReceive field
+	builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors[neighborIndex].ToReceive.Allowed.Prefixes = allowedPrefixes
+
+	return builder
+}
+
+// WithBGPNeighbor defines a single neighbor IP and ASN number.
+func (builder *FrrConfigurationBuilder) WithBGPNeighbor(routerIndex int, bgpPeerIP string,
+	remoteAS uint32) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define BGP Neighbor %s with peer IP %s and ASN %v",
+		builder.Definition.Name, bgpPeerIP, remoteAS)
+
+	if net.ParseIP(bgpPeerIP) == nil {
+		glog.V(100).Infof("The peerIP for the frrConfiguration bgp peer contains invalid ip address %s",
+			bgpPeerIP)
+
+		builder.errorMsg = "frrConfiguration 'peerIP' of the BGPPeer contains invalid ip address"
+	}
+
+	if strconv.Itoa(int(remoteAS)) == "" {
+		glog.V(100).Infof("The remoteASN of the frrConfiguration can not be empty")
+
+		builder.errorMsg = "frrConfiguration 'remoteAS' cannot be empty"
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	if len(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors) == 0 {
+		builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors = []frrtypes.Neighbor{{
+			Address: bgpPeerIP,
+			ASN:     remoteAS,
+		},
+		}
+	} else {
+		builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors =
+			append(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors,
+				frrtypes.Neighbor{
+					Address: bgpPeerIP,
+					ASN:     remoteAS,
+				})
+	}
+
+	return builder
+}
+
+// WithBGPPassword defines the password used between BGP peers to form adjacency and attaches the password using the
+// neighbor index from the neighbor list.
+func (builder *FrrConfigurationBuilder) WithBGPPassword(bgpPassword string,
+	routerIndex, neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define FRRConfiguration %s in namespace %s with the password: %s",
+		builder.Definition.Name, builder.Definition.Namespace, bgpPassword)
+
+	if bgpPassword == "" {
+		glog.V(100).Infof("The bgpPassword should not be blank")
+
+		builder.errorMsg = fmt.Sprintf("the bgpPassword %s is an empty string", bgpPassword)
+	}
+
+	if strconv.Itoa(neighborIndex) == "" {
```

Comment:
Why is this validation only here and not in all other With functions?

Reply:
put on all functions

Reply:
I don't think this validation is necessary, all integers have a string representation that's not blank